File: platform-kotlin-mobile/src/main/java/ru/pocketbyte/locolaser/platform/kotlinmobile/resource/AbsKotlinPlatformResources.java
package ru.pocketbyte.locolaser.platform.kotlinmobile.resource;

import ru.pocketbyte.locolaser.resource.AbsPlatformResources;
import ru.pocketbyte.locolaser.summary.FileSummary;


import java.io.File;

public abstract class AbsKotlinPlatformResources extends AbsPlatformResources {

    public static final String KOTLIN_FILE_EXTENSION = ".kt";

    public final String className;
    public final String classPackage;
    public final String classPackagePath;

    public AbsKotlinPlatformResources(File dir, String name) {
        super(dir, name);

        String[] nameParts = name.split("\\.");

        if (nameParts.length < 2)
            throw new IllegalArgumentException("Invalid resource name");

        this.className = nameParts[nameParts.length - 1];

        StringBuilder packageBuilder = new StringBuilder();
        StringBuilder packagePathBuilder = new StringBuilder();

        for (int i = 0; i < nameParts.length - 1; i++) {
            if (i != 0) {
                packageBuilder.append(".");
                packagePathBuilder.append("/");
            }

            packageBuilder.append(nameParts[i]);
            packagePathBuilder.append(nameParts[i]);
        }

        this.classPackage = packageBuilder.toString();
        this.classPackagePath = packagePathBuilder.toString();
    }

    @Override
    public FileSummary summaryForLocale(String locale) {
        return new FileSummary(getFile());
    }

    protected File getFile() {
        File sourceDir = getDirectory();
        sourceDir.mkdirs();
        return new File(sourceDir, this.classPackagePath + "/" + this.className + KOTLIN_FILE_EXTENSION);
    }
}

File: platform-kotlin-mobile/src/main/java/ru/pocketbyte/locolaser/platform/kotlinmobile/resource/AbsKotlinPlatformResources.kt
package ru.pocketbyte.locolaser.platform.kotlinmobile.resource

import ru.pocketbyte.locolaser.resource.AbsPlatformResources
import ru.pocketbyte.locolaser.summary.FileSummary


import java.io.File

abstract class AbsKotlinPlatformResources(
        dir: File,
        name: String
) : AbsPlatformResources(dir, name) {

    companion object {
        const val KOTLIN_FILE_EXTENSION = ".kt"
    }

    val className: String
    val classPackage: String
    val classPackagePath: String

    protected val file: File
        get() {
            val sourceDir = directory
            sourceDir.mkdirs()
            return File(sourceDir, this.classPackagePath + "/" + this.className + KOTLIN_FILE_EXTENSION)
        }

    init {
        val nameParts = name.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        if (nameParts.size < 2)
            throw IllegalArgumentException("Invalid resource name")

        this.className = nameParts[nameParts.size - 1]

        val packageBuilder = StringBuilder()
        val packagePathBuilder = StringBuilder()

        for (i in 0 until nameParts.size - 1) {
            if (i != 0) {
                packageBuilder.append(".")
                packagePathBuilder.append("/")
            }

            packageBuilder.append(nameParts[i])
            packagePathBuilder.append(nameParts[i])
        }

        this.classPackage = packageBuilder.toString()
        this.classPackagePath = packagePathBuilder.toString()
    }

    override fun summaryForLocale(locale: String): FileSummary {
        return FileSummary(file)
    }
}
